tools/cli/factory.go

```diff
@@ -21,112 +21,87 @@
-	BuildAdminServiceClient(c *cli.Context) (adminserviceclient.Interface, error)
+// ClientFactory is used to construct rpc clients
+type ClientFactory interface {
+	ClientFrontendClient(c *cli.Context) clientFrontend.Interface
+	ServerFrontendClient(c *cli.Context) serverFrontend.Interface
+	ServerAdminClient(c *cli.Context) serverAdmin.Interface
 }
 
-// WorkflowClientBuilder build client to cadence service
-type WorkflowClientBuilder struct {
+type clientFactory struct {
 	hostPort   string
 	dispatcher *yarpc.Dispatcher
 	logger     *zap.Logger
 }
 
-// NewBuilder creates a new WorkflowClientBuilder
-func NewBuilder() *WorkflowClientBuilder {
+// NewClientFactory creates a new ClientFactory
+func NewClientFactory() ClientFactory {
 	logger, err := zap.NewDevelopment()
 	if err != nil {
 		panic(err)
 	}
 
-	return &WorkflowClientBuilder{
+	return &clientFactory{
 		logger: logger,
 	}
 }
 
-// BuildServiceClient builds a rpc service client to cadence service
-func (b *WorkflowClientBuilder) BuildServiceClient(c *cli.Context) (workflowserviceclient.Interface, error) {
-	b.hostPort = localHostPort
-	if addr := c.GlobalString(FlagAddress); addr != "" {
-		b.hostPort = addr
-	}
-
-	if err := b.build(); err != nil {
-		return nil, err
-	}
-
-	if b.dispatcher == nil {
-		b.logger.Fatal("No RPC dispatcher provided to create a connection to Cadence Service")
-	}
-
-	return workflowserviceclient.New(b.dispatcher.ClientConfig(_cadenceFrontendService)), nil
+// ClientFrontendClient builds a frontend client
+func (b *clientFactory) ClientFrontendClient(c *cli.Context) clientFrontend.Interface {
+	b.ensureDispatcher(c)
+	return clientFrontend.New(b.dispatcher.ClientConfig(cadenceFrontendService))
 }
 
-// BuildAdminServiceClient builds a rpc service client to cadence admin service
-func (b *WorkflowClientBuilder) BuildAdminServiceClient(c *cli.Context) (adminserviceclient.Interface, error) {
-	b.hostPort = localHostPort
-	if addr := c.GlobalString(FlagAddress); addr != "" {
-		b.hostPort = addr
-	}
-
-	if err := b.build(); err != nil {
-		return nil, err
-	}
-
-	if b.dispatcher == nil {
-		b.logger.Fatal("No RPC dispatcher provided to create a connection to Cadence Service")
-	}
+// ServerFrontendClient builds a frontend client (based on server side thrift interface)
+func (b *clientFactory) ServerFrontendClient(c *cli.Context) serverFrontend.Interface {
+	b.ensureDispatcher(c)
+	return serverFrontend.New(b.dispatcher.ClientConfig(cadenceFrontendService))
+}
 
-	return adminserviceclient.New(b.dispatcher.ClientConfig(_cadenceFrontendService)), nil
+// ServerAdminClient builds an admin client (based on server side thrift interface)
+func (b *clientFactory) ServerAdminClient(c *cli.Context) serverAdmin.Interface {
+	b.ensureDispatcher(c)
+	return serverAdmin.New(b.dispatcher.ClientConfig(cadenceFrontendService))
 }
 
-func (b *WorkflowClientBuilder) build() error {
+func (b *clientFactory) ensureDispatcher(c *cli.Context) {
```

Comment:
I think we still need to return err and handle in caller. Otherwise these Client may be created with dispatch not running